_posts/2023-07-10-introducing-extensions-for-opensearch.md

```diff
@@ -0,0 +1,127 @@
+
+For implementation details and to examine the differences between Anomaly Detection as a plugin and as an extension, see the [code for Anomaly Detection as an extension](https://github.com/opensearch-project/anomaly-detection/tree/feature/extensions).
+
+## Performance testing
+
+To [benchmark high-cardinality Anomaly Detection (HCAD) as an extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/652), we have replicated the setup described in [this blog post](https://opensearch.org/blog/one-million-enitities-in-one-minute/), replacing 36 r5.2xlarge data nodes with 28 r5.2xlarge data nodes (at the same cost), plus a single r5.16xlarge instance running the Anomaly Detection extension. We analyzed 1,000 historical time periods with 1,000 entities to generate 1 million results:
+
+```json
+GET /_extensions/_ad/detectors/results/_search
+{
+  "query": {
+    "range": {
+      "execution_start_time": {
+        "gte": 1687456200000,
+        "lte": 1687456260000
+      }
+    }
+  },
+  "track_total_hits": true
+}
+```
+
+```json
+{
+    ...
+    "hits": {
+        "total": {
+            "value": 1000000,
+            "relation": "eq"
+        },
+        ...
+    ...
+}
+```
+
+The plugin limits these analysis rates to ensure stable cluster performance. After some fine-tuning (we increased simultaneous detector tasks from 10 to 96 and reduced the batch task piece interval from 5 to 3 seconds), the extension version of Anomaly Detection was able to successfully complete this analysis in 55 seconds. This isn’t quite an apples-to-apples comparison because the test in the previously mentioned blog post involved 1 million smaller models x 1 time slice each to generate 1 million results (we needed more time to implement support for real-time anomaly detection in the extension). However, we believe that we should be able to support 1M models on a single node or on a small number of nodes once extensions are capable of running in that environment.
+
+Because our extension ran on a remote node, we did observe an expected increase in API call latency. For example, the start/stop detector call went from 50--100ms to 200--300ms. These calls are marginal CRUD operations that are used to set up anomaly detection. The following figure provides a latency comparison between extensions and plugins. For more information, see [Performance testing for the Anomaly Detection extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/725#issuecomment-1597874382).
```

Comment:
```suggestion
Because our extension ran on a remote node, we did observe an expected increase in API call latency. For example, the start/stop detector call went from 50--100ms to 200--300ms. These calls are marginal CRUD operations that are used to set up Anomaly Detection. The following figure provides a latency comparison between extensions and plugins. For more information, see [Performance testing for the Anomaly Detection extension](https://github.com/opensearch-project/opensearch-sdk-java/issues/725#issuecomment-1597874382).
```